storage/remote/queue_manager.go

```diff
@@ -1053,6 +1045,11 @@ type sampleOrExemplar struct {
 
 func newQueue(batchSize, capacity int) *queue {
 	batches := capacity / batchSize
+	// Always create an unbuffered channel even if capacity is configured to be
+	// less than max_samples_per_send.
+	if batches == 0 {
+		batches = 1
+	}
```

Comment:
This is the only new change after the rebase. It would probably fix the deadlock in most cases by itself (including the test), but I haven't convinced myself that it would fix all cases so I like the rest of the changes as well. The specific case I am still worried about is if enough samples to completely fill the queue happen to get appended right after the batch_send_deadline expires.